docs/architecture-decision-records/2023-01-switching-to-deno-core.md

```diff
@@ -0,0 +1,126 @@
+
+We decided to take another look at v8 and evaluate it as a possible low-level
+JS/WASM engine for Zinnia.
+
+## Options Considered
+
+### Raw v8
+
+v8 is a C++ project. Fortunately, the [Deno](https://deno.land) project
+maintains a [Rust wrapper](https://crates.io/crates/v8) converting v8's C++ APIs
+into APIs closer to idiomatic Rust.
+
+Unfortunately, working with v8 this way is still too cumbersome.
+
+See e.g. their example code here:
+[v8 example](https://docs.rs/v8/0.62.0/v8/#example)
+
+### `deno_core`
+
+The Deno project is composed from many modules. One of them is `deno_core`,
+which is a library that implements an opinionated way of structuring Rust code
+and exposing it for consumption from JavaScript.
+
+It does not provide any platform APIs like networking and filesystem access.
+
+The host provides a set of Ops (operations) implemented as Rust functions.
+`deno_core` takes care of binding these Rust functions to v8 runtime, exposing
+them to the JavaScript world. The bindings also seamlessly convert between
+JavaScript and Rust types, e.g. `number[]` and `Vec<f64>`.
+
+I wrote a small PoC where the host provides two functions (sync
+`log(msg: string)` and async `sleep(duration: number)`), runs two modules (one
+in JS, one in Rust) and each module invokes both host functions. Source code:
+[runtime-poc/deno-core](https://github.com/filecoin-station/runtime-poc/tree/main/deno-core)
+
+### `deno_runtime`
+
+When using `deno_core`, it's up to us to implement all higher-level
+functionality. What if we could pull the implementation of these features from
+the Deno runtime, instead of implementing them ourselves?
+
+Deno's next building block is `deno_runtime`, which bundles `deno_core` with the
+implementation of different ops and builds a high-level JS API on top of that.
+
+Unfortunately, based on a brief investigation, it seems that `deno_runtime` is
+designed as a self-contained thing that's not open to be extended.
+
+However, the implementations of ops are packages into crates, which are easy to
+add into a `deno_core`-based project.
+
+## Decision
+
+We decided to pivot and build Zinnia on top of `deno_core`.
+
+It will allow us to iterate much faster, while still having a lot of control of
+the API exposed to modules.
+
+While Deno is not perfect, we need to prioritize iteration speed at this stage
+of the project.
+
+## Consequences
+
+- We can support JS & Rust/WASM module from the beginning.
+
+- We can leverage what the Deno team learned about Rust & v8 integration over
+  the years and follow their best practices.
+
+- We can cherry-pick functionality from Deno, either import some of their crates
+  or copy parts of their code (as a last resort solution).
+
+### Caveats
+
+- Supporting both JS and Rust/WASM requires more work. Maybe we should pick one
+  as the first-class option with great documentation & DX, and keep the second
+  one less polished.
```

Comment:
Sgtm! Which are you leaning towards? Or can we create very minimal documentation, and get signals from early module creators?

Reply:
I don't have a strong preference.

On second thought, if we find an easy way how to generate API docs automatically for both the JS & Rust SDKs, then the difference between "first-class" and "less-polished" can be pretty minimal

Reply:
I decided to remove this sentence:

> Maybe we should pick one as the first-class option with great documentation & DX, and keep the second one less polished.